kit/Kit.cpp

```diff
@@ -2888,26 +2916,51 @@ int KitSocketPoll::kitPoll(int timeoutMicroS)
     }
     else
     {
+        const bool loIdleCond = idleHint==LibreOfficeKitIdleHintType::LOK_IDLEHINT_COND_IDLE && timeoutMicroS==0; // LO-Idle Timer (check, conditionally)
+        const bool loIdleLong =
+            idleHint==LibreOfficeKitIdleHintType::LOK_IDLEHINT_IDLE ||
+            (idleHint==LibreOfficeKitIdleHintType::LOK_IDLEHINT_COND_IDLE && timeoutMicroS>2500000); // LO-IdleHandler's SC_IDLE_MAX = 3s, allowing a 500ms lag
         if (checkForIdle)
             timeoutMicroS = 0;
 
-        // Flush at most maxEvents+1, or return when nothing left.
+        // Flush at most maxEvents+1, or return when nothing left,
+        // while looping over time-slices until reaching timeoutMicroS via `poll` + `drainQueue` at 10Hz.
         _pollEnd = startTime + std::chrono::microseconds(timeoutMicroS);
+        int slice = 0;
         do
         {
-            int realTimeout = timeoutMicroS;
-            if (_document && _document->needsQuickPoll())
+            const int timeoutMicroSPre = timeoutMicroS;
+            const size_t tilesPrePoll = _document ? _document->getTileQueueSize() : 0;
+            const bool quickPoll = _document && _document->needsQuickPoll();
+            int realTimeout;
+            if (quickPoll || (slice==0 && (loIdleCond || loIdleLong)))
                 realTimeout = 0;
+            else
+                realTimeout = std::min(100000, timeoutMicroS); // 10Hz 100ms maximum slice
 
-            if (poll(std::chrono::microseconds(realTimeout)) <= 0)
+            const int rc = poll(std::chrono::microseconds(realTimeout));
+            if (rc<0) // poll error
+                break;
+            else if (rc>0) // poll event
+                ++eventsSignalled;
+            else if(realTimeout>0 && timeoutMicroS-realTimeout<=0) // && rc==0: poll timeout on last slice
```

Comment:
Surely we should not have a >0 realTimeout if we have idle work to do anyway - waiting any real time to determine if we are idle doesn't seem helpful.